README.md
<h1 align="center">@cao-mei-you-ren/koishi-plugin-adapter-wechaty </h1>
<p>
  <a href="https://www.npmjs.com/package/@cao-mei-you-ren/koishi-plugin-adapter-wechaty " target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/@cao-mei-you-ren/koishi-plugin-adapter-wechaty.svg">
  </a>
  <a href="https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/actions?query=workflow%3ARelease" target="_blank">
    <img alt="GitHub Workflow Status" src="https://img.shields.io/github/actions/workflow/status/CaoMeiYouRen/koishi-plugin-adapter-wechaty/release.yml?branch=master">
  </a>
  <img src="https://img.shields.io/node/v/@cao-mei-you-ren/koishi-plugin-adapter-wechaty.svg" />
  <a href="https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty#readme" target="_blank">
    <img alt="Documentation" src="https://img.shields.io/badge/documentation-yes-brightgreen.svg" />
  </a>
  <a href="https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/graphs/commit-activity" target="_blank">
    <img alt="Maintenance" src="https://img.shields.io/badge/Maintained%3F-yes-green.svg" />
  </a>
  <a href="https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/blob/master/LICENSE" target="_blank">
    <img alt="License: AGPL-3.0" src="https://img.shields.io/npm/l/@cao-mei-you-ren/koishi-plugin-adapter-wechaty.svg" />
  </a>
</p>

> Koishi Wechaty 适配器
>
> 本项目是原 [koishi-plugin-adapter-wechaty](https://code.mycard.moe/3rdeye/koishi-plugin-adapter-wechaty) 的修复版，在最新的 koishi@4.15.7 可用
>
> Tips: 关于可用性，目前仅在 Windows 下测试通过，Docker(Linux) 平台自测不通过，macOS 缺乏设备。如果有成功接入的用户，请在 issues 或 discussions 中反馈一下，谢谢。

## 🏠 主页

[https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty#readme](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty#readme)


## 📦 依赖要求


- node >=16
- koishi >= 4.15.7

## 🚀 安装

### 插件市场安装

![image-20231207155658055](https://cmyr-picgo.cmyr.ltd/images/202312071557173.png?x-oss-process=style/compressed-picture)

前往插件市场搜索 `@cao-mei-you-ren/adapter-wechaty` 然后安装即可。

**Tips**：被提示为`不安全`是正常的，具体原因可见：[应当避免的情况](https://koishi.chat/zh-CN/cookbook/design/storage.html#%E5%BA%94%E5%BD%93%E9%81%BF%E5%85%8D%E7%9A%84%E6%83%85%E5%86%B5)。

> 简而言之就是 wechaty 这个包本身用到了基于 gyp 的包，故依赖 wechaty 的本项目也被提示 **不安全**

### 手动安装

```sh
npm install @cao-mei-you-ren/koishi-plugin-adapter-wechaty
```

## 👨‍💻 使用

前往插件页面按照引导完成配置，然后点击启用

![image-20231207154722362](https://cmyr-picgo.cmyr.ltd/images/202312071547470.png?x-oss-process=style/compressed-picture)

### name

注意，`name` 字段在启用多个插件实例的时候不可重复！

### puppet

默认为 `wechaty-puppet-wechat`。

详见：https://wechaty.js.org/zh/docs/specs/puppet

### selfId

机器人自身 ID，用于比对是否为机器人自身的消息。

注意和微信号不同。

### ioToken

详见：https://wechaty.js.org/zh/docs/puppet-services/tokens

### wechaty-puppet-wechat

启动插件，并在 Koishi 控制台的『日志』或是 stdout 中查看二维码，扫描登录。

如果一切正常的话，应该可以在日志页面看到如下图片，扫码登录即可

![QQ截图20231207205644](https://cmyr-picgo.cmyr.ltd/images/202312072100520.png?x-oss-process=style/compressed-picture)

## 🛠️ 开发

```sh
npm run dev
```

## 🔧 编译

```sh
npm run build
```

## 🔍 Lint

```sh
npm run lint
```

## 💾 Commit

```sh
npm run commit
```


## 👤 作者

**CaoMeiYouRen**

* Website: [https://blog.cmyr.ltd/](https://blog.cmyr.ltd/)

* GitHub: [@CaoMeiYouRen](https://github.com/CaoMeiYouRen)

## 🤝 贡献/提问

欢迎 贡献、提问或提出新功能！<br />如有问题请查看 [issues page](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/issues). <br/>贡献或提出新功能可以查看[contributing guide](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/blob/master/CONTRIBUTING.md).

### 关于问题反馈的额外说明

可以向本项目提出问题和反馈，但不一定能得到解决。

由于本项目的核心代码都是从[原项目](https://code.mycard.moe/3rdeye/koishi-plugin-adapter-wechaty)复制的，所以很多逻辑我也没有理清楚，故仅提供基础维护，即：确保在 `Wechaty`本身可用的情况下确保`@cao-mei-you-ren/koishi-plugin-adapter-wechaty`可用。

### 关于 License

由于原项目使用了 [AGPL-3.0](https://www.gnu.org/licenses/agpl-3.0.zh-cn.html)，故根据 AGPL-3.0，本项目也需要使用 AGPL-3.0 进行开源。

## 💰 支持

如果觉得这个项目有用的话请给一颗⭐️，非常感谢

<a href="https://afdian.net/@CaoMeiYouRen">
  <img src="https://cdn.jsdelivr.net/gh/CaoMeiYouRen/image-hosting-01@master/images/202306192324870.png" width="312px" height="78px" alt="在爱发电支持我">
</a>

<a href="https://patreon.com/CaoMeiYouRen">
    <img src="https://cdn.jsdelivr.net/gh/CaoMeiYouRen/image-hosting-01@master/images/202306142054108.svg" width="312px" height="78px" alt="become a patreon"/>
</a>

## 🌟 Star History

[![Star History Chart](https://api.star-history.com/svg?repos=CaoMeiYouRen/koishi-plugin-adapter-wechaty&type=Date)](https://star-history.com/#CaoMeiYouRen/koishi-plugin-adapter-wechaty&Date)

## 📝 License

Copyright © 2023 [CaoMeiYouRen](https://github.com/CaoMeiYouRen).<br />
This project is [AGPL-3.0](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/blob/master/LICENSE) licensed.

***
_This README was generated with ❤️ by [cmyr-template-cli](https://github.com/CaoMeiYouRen/cmyr-template-cli)_


CHANGELOG.md
# @cao-mei-you-ren/koishi-plugin-adapter-wechaty

## [1.0.2](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/compare/v1.0.1...v1.0.2) (2023-12-07)


### 🐛 Bug 修复

* 修复 头像的跨域问题 ([d97ee88](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/commit/d97ee88))

## [1.0.1](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/compare/v1.0.0...v1.0.1) (2023-12-07)


### 🐛 Bug 修复

* 优化 dist 和 types 导出 ([ebb523c](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/commit/ebb523c))

# 1.0.0 (2023-12-07)


### ✨ 新功能

* 完成项目发布预备 ([60bfd01](https://github.com/CaoMeiYouRen/koishi-plugin-adapter-wechaty/commit/60bfd01))


src/type.ts
import { WechatyBuilder } from 'wechaty'
import type { WechatyEventListeners } from 'wechaty/src/schemas/mod'
import { Session } from 'koishi'
import WechatyBot from './index'

declare module 'koishi' {
    interface Events extends WechatyEvents { }
}

export type WechatyInstance = ReturnType<(typeof WechatyBuilder)['build']>

type AddSessionToFront<F> = F extends (...args: infer A) => infer P
    ? (session: Session & { bot: WechatyBot }, ...args: A) => P
    : never
export const PassthroughEvents = [
    'scan',
    'room-topic',
    'heartbeat',
    'error',
] as const
// eslint-disable-next-line @typescript-eslint/no-redeclare
export type PassthroughEvents = (typeof PassthroughEvents)[number]
export type WechatyEvents = {
    [K in keyof WechatyEventListeners as `wechaty/${PassthroughEvents}`]: AddSessionToFront<
    WechatyEventListeners[K]
    >;
}


src/utils.ts
import path from 'path'
import { segment, Universal, Element, Awaitable, Logger } from 'koishi'
import type { ContactInterface } from 'wechaty/src/user-modules/contact'
import type { RoomInterface } from 'wechaty/src/user-modules/room'
import type { MessageInterface } from 'wechaty/src/user-modules/message'
import FileType from 'file-type'
import mime2ext from 'mime2ext'
import { FileBox } from 'file-box'
import WechatyBot from './index'

export type ContactLike = Pick<
ContactInterface,
'id' | 'name' | 'avatar' | 'self'
>
export type FileBoxLike = Awaited<ReturnType<ContactLike['avatar']>>

export type RoomLike = Pick<RoomInterface, 'id' | 'topic'>
export type MessageLike = MessageInterface

export const fileBoxToUrl = async (file: FileBoxLike): Promise<string> => {
    if (!file) {
        return undefined
    }
    if (file['remoteUrl']) {
        return file['remoteUrl']
    }
    let buf: Buffer
    try {
        buf = await file.toBuffer()
    } catch (e) {
        buf = file['stream']
    }
    const fileType = await FileType.fromBuffer(buf)
    const mime = fileType ? fileType.mime : 'application/octet-stream'
    return `data:${mime};base64,${buf.toString('base64')}`
}

export const adaptContact = async (
    contact: ContactLike,
): Promise<Universal.User> => ({
    id: contact.id,
    userId: contact.id,
    nickname: contact.name(),
    avatar: await fileBoxToUrl(await contact.avatar()), // await (await contact.avatar()).toDataURL(),
    isBot: contact.self(),
})

export const adaptRoom = async (
    room: RoomLike,
): Promise<Universal.Channel & Universal.Guild> => {
    const name = await room.topic()

    return {
        id: room.id,
        name,
        type: 0,
    }
}

async function extractMessageURL(
    message: MessageLike,
    segmentFactory: (url: string, name: string) => Awaitable<Element>,
): Promise<Element> {
    const file = await message.toFileBox()
    if (!file) {
        return
    }
    return segmentFactory(await fileBoxToUrl(file), file.name)
}

export async function messageToElement(
    bot: WechatyBot,
    message: MessageLike,
): Promise<Element[]> {
    try {
        const MessageType = bot.internal.Message.Type
        const elements: Element[] = []
        const mentions = await message.mentionList()
        switch (message.type()) {
            case MessageType.Recalled:
                return []
            case MessageType.Text:{
                let text = message.text()
                for (const mention of mentions) {
                    const name = mention.name()
                    text = text.replace(new RegExp(`@${name}\\s+`, 'g'), '')
                }
                elements.push(segment.text(text))
                break
            }
            case MessageType.Image:
                elements.push(
                    await extractMessageURL(message, async (url, name) => segment.image(url, { file: await autoFilename(url) }),
                    ),
                )
                break
            case MessageType.Audio:
                elements.push(
                    await extractMessageURL(message, async (url, name) => segment.audio(url, { file: await autoFilename(url) }),
                    ),
                )
                break
            case MessageType.Video:
                elements.push(
                    await extractMessageURL(message, async (url, name) => segment.video(url, { file: await autoFilename(url) }),
                    ),
                )
                break
            case MessageType.Attachment:
                elements.push(
                    await extractMessageURL(message, async (url, name) => segment.file(url, { file: name }),
                    ),
                )
                break
            case MessageType.Url:{
                const link = await message.toUrlLink()
                elements.push(
                    segment('a', { href: link.url() }, [
                        `${link.title()}\n${link.description}`,
                        segment.image(link.thumbnailUrl()),
                    ]),
                )
                break
            }
            case MessageType.Contact:{
                const contact = await message.toContact()
                elements.push(
                    segment('wechaty:contact', { id: contact.id, name: contact.name() }),
                )
                break
            }
            default:
                return
        }
        mentions.forEach((mention) => elements.unshift(segment.at(mention.id)))
        return elements
    } catch (e) {
        console.error(e)
    }
}

export async function adaptMessage(
    bot: WechatyBot,
    message: MessageLike,
): Promise<Universal.Message> {
    const elements = await messageToElement(bot, message)
    if (!elements) {
        return
    }
    const room = message.room()
    const from = message.talker()
    if (!from) {
        return
    }
    const user = await adaptContact(from as any)
    const channel = room ? await adaptRoom(room as any) : {
        id: `private:${user.id}`,
        type: 0,
    }
    return {
        id: message.id,
        messageId: message.id,
        user,
        channel,
        guild: channel,
        elements: elements as unknown as Universal.Message['elements'],
        content: elements.map((e) => e.toString()).join(''),
        timestamp: (message.date() || new Date()).valueOf(),
        createdAt: (message.date() || new Date()).valueOf(),
    }
}

export async function autoFilename(url: string) {
    if (url.startsWith('file://')) {
        return path.basename(url.slice(7))
    }
    if (url.startsWith('base64://')) {
        const buf = Buffer.from(url.slice(9), 'base64')
        const type = await FileType.fromBuffer(buf)
        return `file.${type.ext}`
    }
    if (url.startsWith('data:')) {
        const [, mime, base64] = url.match(/^data:([^;]+);base64,(.+)$/)
        const ext = mime2ext(mime)
        if (ext) {
            return `file.${ext}`
        }
        const buf = Buffer.from(base64, 'base64')
        const type = await FileType.fromBuffer(buf)
        return `file.${type?.ext || 'bin'}`
    }
    return path.basename(new URL(url).pathname)
}

export const elementToFileBox = async (element: Element) => {
    const { attrs } = element
    const { url, file } = attrs
    if (!url) {
        return
    }
    if (url.startsWith('file://')) {
        const filePath = url.slice(7)
        return FileBox.fromFile(filePath, file || await autoFilename(url))
    }
    if (url.startsWith('base64://')) {
        return FileBox.fromBase64(url.slice(9), file || await autoFilename(url))
    }
    if (url.startsWith('data:')) {
        const [, mime, base64] = url.match(/^data:([^;]+);base64,(.+)$/)
        const ext = mime2ext(mime) || 'bin'
        return FileBox.fromBase64(base64, file || `file.${ext}`)
    }
    return FileBox.fromUrl(url, {
        name: file || await autoFilename(url),
    })
}


src/adapter.ts
import { Adapter, Awaitable, Context, Logger, MaybeArray, Session, makeArray } from 'koishi'
import type { WechatyEventListeners } from 'wechaty/src/schemas/mod'
import type { ContactInterface } from 'wechaty/src/user-modules/contact'
import { WechatyBot } from './bot'
import { PassthroughEvents } from './type'
import { adaptContact, adaptMessage, adaptRoom } from './utils'

// 接受消息
export default class WechatyAdapter<C extends Context = Context> extends Adapter<C, WechatyBot<C>> {

    private readonly logger = new Logger(WechatyAdapter.name)

    async connect(bot: WechatyBot<C>) {
        this.logger.debug('connect')
        bot.initialize()
        this.loadEvents(bot)
        this.emitWechatyEvent(bot)
        await bot.internal.start()
        bot.online()
        bot.loginTime = Date.now()
    }

    async disconnect(bot: WechatyBot<C>) {
        this.logger.debug('disconnect')
        await bot.internal.stop()
        bot.offline()
    }

    private adaptEvent<K extends keyof WechatyEventListeners>(
        bot: WechatyBot<C>,
        event: K,
        sessionBuilder: (
            ...args: Parameters<WechatyEventListeners[K]>
        ) => Awaitable<MaybeArray<Partial<Session>>>) {
        bot.internal.on(event, async (...args: any[]) => {
            const timestamp = Date.now()
            const payloads = makeArray(
                await sessionBuilder(...(args as Parameters<WechatyEventListeners[K]>)),
            )
            payloads.forEach(async (payload) => {
                if (!payload) {
                    return
                }
                payload.timestamp ??= timestamp
                // this.logger.debug('payload', payload)
                const session = bot.session(payload as any)
                // Object.assign(session, {
                //   ...payload
                // })
                if (payload.content && !session.content) {
                    session.content = payload.content
                }
                if (payload.messageId && !session.messageId) {
                    session.messageId = payload.messageId
                }
                this.logger.debug('session', session)
                if (
                    session?.type === 'message' &&
          (session.timestamp ?? Date.now()) - bot.loginTime < 1000 * 2 // 弃用机器人启动之前的消息，避免刷屏
                ) {
                    // this.logger.debug('session 已过时, content: %s', session.content)
                    return
                }
                bot.dispatch(session)
            })
        })
    }

    private loadEvents(bot: WechatyBot<C>) {
        this.adaptEvent(bot, 'message', async (message) => {
            this.logger.debug('message', message)
            const adaptedMessage = await adaptMessage(bot, message)
            if (!adaptedMessage) {
                return
            }
            let messageType = 'message'
            if ( message.type() === bot.internal.Message.Type.Recalled) {
                messageType = 'message-deleted'
            } else if (adaptedMessage.user.id === bot.selfId) {
                messageType = 'message-sent'
            }
            return {
                ...adaptedMessage,
                type: messageType,
            } as any
        })
        this.adaptEvent(bot, 'friendship', async (friendship) => {
            const wechatyType = friendship.type()
            let type: string
            if (wechatyType === 2) {
                // receive
                type = 'friend-request'
            } else if (wechatyType === 1) {
                // confirm
                type = 'friend-added'
            } else {
                return
            }
            const channelId = `private:${friendship.contact().id}`
            return {
                type,
                id: friendship.id,
                messageId: friendship.id,
                content: friendship.hello(),
                channelId,
                channel: {
                    id: channelId,
                },
                user: await adaptContact(friendship.contact()),
            } as any
        })
        this.adaptEvent(bot, 'room-invite', async (invitation) => {
            const channelName = await invitation.topic()
            return {
                type: 'guild-request',
                id: invitation.id,
                messageId: invitation.id,
                user: await adaptContact(await invitation.inviter()),
                channel: {
                    id: channelName,
                    name: channelName,
                },
                channelName,
                channelId: channelName,
                guild: {
                    id: channelName,
                    name: channelName,
                },
                guildName: channelName,
                guildId: channelName,
            } as any
        })
        this.adaptEvent(
            bot,
            'room-join',
            async (
                room,
                inviteeList: ContactInterface[],
                inviter: ContactInterface,
                date: Date,
            ) => {
                const channel = await adaptRoom(room)
                const timestamp = (date || new Date()).valueOf()
                const sessions = await Promise.all(
                    inviteeList.map(async (invitee): Promise<Partial<Session>> => {
                        const invitedUser = await adaptContact(invitee)
                        return {
                            type: invitee.self() ? 'guild-added' : 'guild-member-added',
                            subtype: invitedUser.id === inviter.id ? 'active' : 'passive',
                            ...invitedUser,
                            targetId: invitedUser.id,
                        } as any
                    }),
                )
                return sessions.map((s) => ({
                    operatorId: inviter.id,
                    timestamp,
                    ...channel,
                    ...s,
                })) as any
            },
        )
        this.adaptEvent(
            bot,
            'room-leave',
            async (
                room,
                leaverList: ContactInterface[],
                operator: ContactInterface,
                date: Date,
            ) => {
                const channel = await adaptRoom(room)
                const timestamp = (date || new Date()).valueOf()
                const sessions = await Promise.all(
                    leaverList.map(async (leaver): Promise<Partial<Session>> => {
                        const user = await adaptContact(leaver)
                        return {
                            type: leaver.self() ? 'guild-deleted' : 'guild-member-deleted',
                            subtype:
                !operator || user.id === operator.id ? 'active' : 'passive',
                            targetId: user.id,
                            ...user,
                        } as any
                    }),
                )
                return sessions.map((s) => ({
                    ...channel,
                    operatorId: operator?.id,
                    timestamp,
                    ...s,
                })) as any
            },
        )
    }

    /**
   * 向全局广播 wechaty/* 事件
   *
   * @author CaoMeiYouRen
   * @date 2023-12-06
   * @private
   */
    private emitWechatyEvent(bot: WechatyBot<C>) {
        for (const event of PassthroughEvents) {
            bot.internal.on(event, (...args: any) => {
                const session = bot.session()
                session.type = `wechaty/${event}`
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                session.app.emit(session, `wechaty/${event}`, session, ...args)
            })
        }
    }
}


src/message.ts
import { Context, Element, MessageEncoder, segment, Session } from 'koishi'
import { Contact, Sayable } from 'wechaty'
import { WechatyBot } from './bot'
import { elementToFileBox } from './utils'

export class WechatyMessenger<C extends Context = Context> extends MessageEncoder<C, WechatyBot<C>> {
    buffer = ''
    contactCache = new Map<string, Contact>()

    addMessage(messageId: string) {
        if (!messageId) {
            return
        }
        const session = this.bot.session() as Session
        session.messageId = messageId
        session.app.emit(session, 'send', session)
        this.results.push({
            ...session,
            id: messageId,
        } as any)
    }

    private isGuild() {
        return !this.channelId.startsWith('private:')
    }

    async post(content: Sayable) {
        try {
            if (!this.isGuild()) {
                const contact = await this.bot.internal.Contact.find({
                    id: this.channelId.slice(8),
                })
                if (!contact) {
                    return
                }
                const message = await contact.say(content)
                if (!message) {
                    return
                }
                this.addMessage(message.id)
            } else {
                const room = await this.bot.internal.Room.find({ id: this.channelId })
                if (!room) {
                    return
                }
                const message = await room.say(content)
                if (!message) {
                    return
                }
                this.addMessage(message.id)
            }
        } catch (e) {
            this.errors.push(e)
        }
    }

    async flush() {
        if (!this.buffer) {
            return
        }
        await this.post(this.buffer)
        this.buffer = ''
    }

    async sendMedia(media: Element) {
        const fileBox = await elementToFileBox(media)
        if (!fileBox) {
            return
        }
        return this.post(fileBox as any)
    }

    private text(content: string) {
        this.buffer += content
    }

    private async getRelatedContact(id: string) {
        if (this.contactCache.has(id)) {
            return this.contactCache.get(id)
        }
        let member: Contact
        if (this.isGuild()) {
            const room = await this.bot.internal.Room.find({ id: this.channelId })
            if (!room) {
                return
            }
            const members = await room.memberAll()
            // eslint-disable-next-line @typescript-eslint/no-shadow
            member = members.find((member) => member.id === id)
        } else {
            member = await this.bot.internal.Contact.find({ id })
        }
        if (!member) {
            return
        }
        this.contactCache.set(id, member)
        return member
    }

    async visit(element: segment) {
        const { type, attrs, children } = element
        switch (type) {
            case 'text':
                this.text(attrs.content)
                break
            case 'p':
                await this.render(children)
                this.text('\n')
                break
            case 'a':
                await this.render(children)
                if (attrs.href) {
                    this.text(` (${attrs.href}) `)
                }
                break
            case 'at':
                if (attrs.id) {
                    const contact = await this.getRelatedContact(attrs.id)
                    if (contact) {
                        this.text(`@${contact.name()} `)
                    }
                } else if (attrs.type === 'all') {
                    this.text('@全体成员 ')
                } else if (attrs.type === 'here') {
                    this.text('@在线成员 ')
                } else if (attrs.role) {
                    this.text(`@${attrs.role}`)
                }
                break
            case 'image':
            case 'video':
            case 'audio':
            case 'file':
                await this.flush()
                await this.sendMedia(element)
                break
            case 'figure':
            case 'message':
                await this.flush()
                await this.render(children, true)
                break
            case 'wechaty:contact':{
                const contact = await this.getRelatedContact(attrs.id)
                if (contact) {
                    await this.flush()
                    await this.post(contact)
                }
                break
            }
            default:
                await this.render(children)
        }
    }
}


src/bot.ts
import os from 'os'
import { Bot, Context, Fragment, Logger, Schema, Random } from 'koishi'
import { WechatyBuilder, WechatyOptions } from 'wechaty'
import qrcodeTerminal from 'qrcode-terminal'
import type { SendOptions } from '@satorijs/protocol'
import { WechatyInstance } from './type'
import WechatyAdapter from './adapter'
import { adaptMessage, adaptContact, adaptRoom, fileBoxToUrl } from './utils'
import { WechatyMessenger } from './message'

// 发送消息
export class WechatyBot<C extends Context = Context> extends Bot<C, WechatyBot.Config> {

    // static MessageEncoder = WechatyMessenger

    readonly logger = new Logger(WechatyBot.name)

    readonly internal: WechatyInstance

    loginTime: number

    constructor(ctx: C, config: WechatyBot.Config) {
        super(ctx, config)
        this.platform = 'wechaty'
        // this.logger.debug(this.config)
        this.internal = WechatyBuilder.build(this.config)
        ctx.plugin(WechatyAdapter, this)
    }

    // async start() {
    //   this.logger.debug('start')
    //   this.init()
    //   await this.internal.start()
    //   this.online()
    // }

    // async stop() {
    //   this.logger.debug('stop')
    //   await this.internal.stop()
    //   this.offline()
    // }

    async sendMessage(
        channelId: string,
        content: Fragment,
        guildId?: string,
        options?: SendOptions,
    ) {
        return (await new WechatyMessenger(this, channelId, guildId, options)
            .send(content)).map((e) => e.id)
    }

    async sendPrivateMessage(
        userId: string,
        content: Fragment,
        guildId?: string,
        options?: SendOptions,
    ) {
        return (await new WechatyMessenger(
            this,
            `private:${userId}`,
            guildId,
            options,
        ).send(content)).map((e) => e.id)
    }
    async getMessage(channelId: string, messageId: string) {
        const message = await this.internal.Message.find({ id: messageId })
        if (!message) {
            return
        }
        return adaptMessage(this, message as any)
    }
    async getMessageList(channelId: string, next?: string) {
        const messages = await this.internal.Message.findAll({
            roomId: !channelId.startsWith('private:') ? channelId : undefined,
            fromId: channelId.startsWith('private:') ? channelId.slice(8) : undefined,
        })
        return {
            data: await Promise.all(messages.map((m) => adaptMessage(this, m as any))),
            next,
        }
    }
    async editMessage(channelId: string, messageId: string, content: any) {
    }

    async deleteMessage(channelId: string, messageId: string) {
    }

    // user
    async getSelf() {
        const self = this.internal.currentUser
        return adaptContact(self)
    }
    async getUser(userId: string) {
        const contact = await this.internal.Contact.find({ id: userId })
        return adaptContact(contact)
    }
    async getFriendList() {
        const friends = await this.internal.Contact.findAll()
        return {
            data: await Promise.all(friends.map(adaptContact)),
        }
    }
    async deleteFriend(userId: string) { }

    // guild
    async getGuild(guildId: string) {
        const room = await this.internal.Room.find({ id: guildId })
        return adaptRoom(room)
    }

    async getGuildList() {
        const rooms = await this.internal.Room.findAll()
        return {
            data: await Promise.all(rooms.map(adaptRoom)),
        }
    }

    async getChannel(channelId: string) {
        return this.getGuild(channelId)
    }

    async getChannelList() {
        return this.getGuildList()
    }

    async muteChannel(channelId: string, guildId?: string, enable?: boolean) { }

    // guild member
    async getGuildMember(guildId: string, userId: string) {
        const room = await this.internal.Room.find({ id: guildId })
        if (!room) {
            return
        }
        const members = await room.memberAll()
        const member = members.find((m) => m.id === userId)
        if (!member) {
            return
        }
        return adaptContact(member)
    }

    async getGuildMemberList(guildId: string) {
        const room = await this.internal.Room.find({ id: guildId })
        if (!room) {
            return
        }
        const members = await room.memberAll()
        return {
            data: await Promise.all(members.map(adaptContact)),
        }
    }
    async kickGuildMember(guildId: string, userId: string, permanent?: boolean) { }

    async muteGuildMember(
        guildId: string,
        userId: string,
        duration: number,
        reason?: string,
    ) { }

    // request
    async handleFriendRequest(
        messageId: string,
        approve: boolean,
        comment?: string,
    ) {
        if (!approve) {
            return
        }
        return this.internal.Friendship.load(messageId).accept()
    }

    async handleGuildRequest(
        messageId: string,
        approve: boolean,
        comment?: string,
    ) {
        if (!approve) {
            return
        }
        return this.internal.RoomInvitation.load(messageId).accept()
    }

    async handleGuildMemberRequest(
        messageId: string,
        approve: boolean,
        comment?: string,
    ) { }

    /**
   * 初始化事件监听
   *
   * @author CaoMeiYouRen
   * @date 2023-12-06
   * @private
   */
    initialize() {
        this.loginTime = Date.now()
        this.internal.on('login', async (user) => {
            this.logger.debug('user', user)
            this.selfId = user.id
            this.user = this.user || ({} as any)
            this.user.name = user.name()
            this.user.avatar = await (await user.avatar()).toDataURL() // await fileBoxToUrl(await user.avatar()) //
            this.online()
        })
        this.internal.on('logout', () => {
            this.offline()
        })
        this.internal.on('error', (error) => {
            this.offline(error)
        })
        this.internal.on('scan', (qrcode, status) => {
            qrcodeTerminal.generate(qrcode, { small: false }, (img) => this.logger?.info(
                `Scan (${status}): https://wechaty.js.org/qrcode/${encodeURIComponent(
                    qrcode,
                )}\n${img}`,
            ),
            )
        })

    }

}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace WechatyBot {
    export interface Config {
        name: string
        puppet: WechatyOptions['puppet']
        puppetOptions?: any
        selfId?: string
        ioToken?: string
    }

    export const Config: Schema<Config> = Schema.object({
        name: Schema.string().required().default(`wechaty-${Random.int(0, 1000).toFixed(0).padStart(4, '0')}`).description('Wechaty 配置文件保存名称。启用多个插件实例时不可重复'),
        puppet: Schema.union(['wechaty-puppet-wechat']).default('wechaty-puppet-wechat').description('Wechaty 使用的 Puppet。').hidden(os.platform() !== 'win32') as Schema<WechatyOptions['puppet']>,
        puppetOptions: Schema.object({ uos: Schema.boolean().default(true) }).description('Wechaty Puppet 选项。').hidden(),
        selfId: Schema.string().description('机器人自身 ID'),
        ioToken: Schema.string().description('Io TOKEN，详见 https://wechaty.js.org/docs/puppet-services/tokens?_highlight=token#get-a-token'),
    })
}
